# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd
from typing import Union

file_data_path = "data/supermarket_sales.csv"
supermarket_sales_df = pd.read_csv(file_data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [4]:
display(supermarket_sales_df.head(4), supermarket_sales_df.tail(4))

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [5]:
supermarket_sales_df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [6]:
supermarket_sales_df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [7]:
supermarket_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [3]:
supermarket_sales_df.describe(percentiles=[.01, .25, .5, .75, .99])['gross income']

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

<p style='font-family: Garamond ; font-size: 16px;'>75-й персентиль означає, що 75% всіх значень "gross income" (валового доходу) менші або дорівнюють 22.45, а решта 25% — більші за це значення.</p>
<p style='font-family: Garamond ; font-size: 16px;'>Ми взяли 1000 інвойсів і впорядкували їхній валовий дохід (gross income) від найменшого до найбільшого.75-й персентиль показує, який дохід маємо від інвойсу, що стоїть на 750-й позиції у цьому списку. Тобто у більшості випадків (три чверті інвойсів) валовий дохід буде меншим або рівним 22.45, і лише від чверті інвойсів матимемо дохід більший за цю суму.</p>

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [9]:
city_counts = supermarket_sales_df["City"].value_counts()
city_counts.name = "City"
print(city_counts)

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: City, dtype: int64


6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [11]:
city_counts/city_counts.sum() * 100

City
Yangon       34.0
Mandalay     33.2
Naypyitaw    32.8
Name: City, dtype: float64

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [3]:
# Filter data for cities Yangon and Mandalay
filtered_data = supermarket_sales_df[supermarket_sales_df['City'].isin(['Yangon', 'Mandalay'])]

# Calculate the average gross income for each city
average_gross_income = filtered_data.groupby('City')['gross income'].mean()

average_gross_income

City
Mandalay    15.232024
Yangon      14.874001
Name: gross income, dtype: float64

In [4]:
filtered_data.head()

Invoice ID Branch    City Customer type  Gender            Product line  \
0  750-67-8428      A  Yangon        Member  Female       Health and beauty   
2  631-41-3108      A  Yangon        Normal    Male      Home and lifestyle   
3  123-19-1176      A  Yangon        Member    Male       Health and beauty   
4  373-73-7910      A  Yangon        Normal    Male       Sports and travel   
6  355-53-5943      A  Yangon        Member  Female  Electronic accessories   

   Unit price  Quantity   Tax 5%     Total       Date   Time      Payment  \
0       74.69         7  26.1415  548.9715   1/5/2019  13:08      Ewallet   
2       46.33         7  16.2155  340.5255   3/3/2019  13:23  Credit card   
3       58.22         8  23.2880  489.0480  1/27/2019  20:33      Ewallet   
4       86.31         7  30.2085  634.3785   2/8/2019  10:37      Ewallet   
6       68.84         6  20.6520  433.6920  2/25/2019  14:36      Ewallet   

     cogs  gross margin percentage  gross income  Rating  
0  522.83                 4.761905       26.1415     9.1  
2  324.31                 4.761905       16.2155     7.4  
3  465.76                 4.761905       23.2880     8.4  
4  604.17                 4.761905       30.2085     5.3  
6  413.04                 4.761905       20.6520     5.8

<h1 style='font-family: Garamond ;'>Гіпотеза 1: Вища частка покупок дорогих товарів у <b>Mandalay</b></h1>
<hr></hr>
<p style='font-family: Garamond ; font-size: 18px;'>У Mandalay, можливо, клієнти частіше купують товари з вищою ціною (наприклад, електроніку чи товари для дому). Це впливає на підвищення показника gross income, оскільки ці товари мають більшу маржу.</p>

In [13]:
product_analysis = filtered_data.groupby(['City', 'Product line']).agg({
    'Total': ['mean', 'sum', 'count']
}).reset_index()

product_analysis.columns = ['City', 'Product line', 'Average Total', 'Total Revenue', 'Transaction Count']

mandalay_products = product_analysis[product_analysis['City'] == 'Mandalay']
yangon_products = product_analysis[product_analysis['City'] == 'Yangon']

print('\nMandalay. Product line distribution analysis:\n')
display(mandalay_products, 
        mandalay_products[['Product line','Average Total']].agg(["min", "max"]),
        mandalay_products[['Product line','Total Revenue']].agg(["min", "max"]))


Mandalay. Product line distribution analysis:



City            Product line  Average Total  Total Revenue  \
0  Mandalay  Electronic accessories     310.026245     17051.4435   
1  Mandalay     Fashion accessories     264.730911     16413.3165   
2  Mandalay      Food and beverages     304.297770     15214.8885   
3  Mandalay       Health and beauty     376.993585     19980.6600   
4  Mandalay      Home and lifestyle     350.983290     17549.1645   
5  Mandalay       Sports and travel     322.390306     19988.1990   

   Transaction Count  
0                 55  
1                 62  
2                 50  
3                 53  
4                 50  
5                 62

Product line  Average Total
min  Electronic accessories     264.730911
max       Sports and travel     376.993585

Product line  Total Revenue
min  Electronic accessories     15214.8885
max       Sports and travel     19988.1990

In [14]:
print('\nYangon. Product line distribution analysis:\n')
display(yangon_products, 
        yangon_products[['Product line','Average Total']].agg(["min", "max"]),
        yangon_products[['Product line','Total Revenue']].agg(["min", "max"]))


Yangon. Product line distribution analysis:



City            Product line  Average Total  Total Revenue  \
6   Yangon  Electronic accessories     305.285225     18317.1135   
7   Yangon     Fashion accessories     320.245265     16332.5085   
8   Yangon      Food and beverages     295.915526     17163.1005   
9   Yangon       Health and beauty     268.037298     12597.7530   
10  Yangon      Home and lifestyle     344.879931     22417.1955   
11  Yangon       Sports and travel     328.350839     19372.6995   

    Transaction Count  
6                  60  
7                  51  
8                  58  
9                  47  
10                 65  
11                 59

Product line  Average Total
min  Electronic accessories     268.037298
max       Sports and travel     344.879931

Product line  Total Revenue
min  Electronic accessories     12597.7530
max       Sports and travel     22417.1955

<ol style='font-family: Garamond ; font-size: 16px;'>
    <li>Середній чек у Mandalay трохи вищий:</li>
        <ul ol style='font-family: Garamond ; font-size: 16px;'>
            <li>Найдорожчий продукт у Mandalay має середній чек 376.99, тоді як у Yangon — 344.88.</li>
            <li>Також у Mandalay більше продуктів з середнім чеком вище 300 (5 з 6), порівняно з Yangon (4 з 6).</li>
        </ul>
    <li>Категорії з вищими середніми чеками в Mandalay включають (у Yangon ці ж категорії мають нижчі показники.):</li>
        <ul style='font-family: Garamond ; font-size: 16px;'>
            <li>Health and beauty (376.99).</li>
            <li>Home and lifestyle (350.98).</li>
            <li>Electronic accessories (310.03)</li>
            <li>Food and beverages (304.30).</li>
        </ul>
    <li>Total Revenue у Mandalay та Yangon не дає перевагу Mandalay:</li>
        <ul style='font-family: Garamond ; font-size: 16px;'>
            <li>Максимальна загальна виручка в Yangon навіть вища (22417.19) завдяки Home and lifestyle.</li>
            <li>Це свідчить, що кількість транзакцій і сукупна виручка подібні або навіть вищі в Yangon, але середній чек все ж вищий у Mandalay.</li>
        </ul>
</ol>    
<p style='font-family: Garamond ; font-size: 16px;'>У Mandalay дійсно частіше купуються товари з вищим середнім чеком, особливо в категоріях Health and beauty, Home and lifestyle та Electronic accessories. Це може свідчити про більший попит на дорогі товари, що впливає на підвищення показника gross income.</p>
<p style='font-family: Garamond ; font-size: 16px;'>Однак загальна виручка (Total Revenue) у Yangon за деякими категоріями навіть перевищує Mandalay, що означає, що кількість покупок теж відіграє важливу роль.</p>

<h1 style='font-family: Garamond ;'>Гіпотеза 2: Вищий рівень клієнтів-членів (клієнтів-власників карток/учасників програм постійних клієнтів) <code>Members</code> <b>Mandalay</b></h1>
<hr></hr>
<p style='font-family: Garamond ; font-size: 18px;'>Члени програми лояльності (<code>Customer type</code>) можуть бути схильні витрачати більше завдяки знижкам або спеціальним пропозиціям.</p>

In [17]:
customer_type_analysis = filtered_data.groupby(['City', 'Customer type']).agg({
    'Total': ['mean', 'sum', 'count']
}).reset_index()

customer_type_analysis.columns = ['City', 'Customer Type', 'Average Total', 'Total Revenue', 'Transaction Count']

mandalay_customer_type = customer_type_analysis[customer_type_analysis['City'] == 'Mandalay']
yangon_customer_type = customer_type_analysis[customer_type_analysis['City'] == 'Yangon']

display(mandalay_customer_type, yangon_customer_type)

City Customer Type  Average Total  Total Revenue  Transaction Count
0  Mandalay        Member     325.482945      53704.686                165
1  Mandalay        Normal     314.329257      52492.986                167

City Customer Type  Average Total  Total Revenue  Transaction Count
2  Yangon        Member     321.182488     53637.4755                167
3  Yangon        Normal     303.831763     52562.8950                173

<ol style='font-family: Garamond ; font-size: 16px;'>
    <li>У Mandalay середній чек для Members вищий (325.48) ніж для Normal (314.33):</li>
        <ul ol style='font-family: Garamond ; font-size: 16px;'>
            <li>Різниця приблизно 11.15 одиниць на одну покупку.</li>
            <li>Схожа ситуація спостерігається і в Yangon (різниця ~17.35 одиниць).</li>
        </ul>
    <li>Загальна виручка (Total Revenue) теж вища для Members у Mandalay (53704.69) порівняно з Normal (52492.99), незважаючи на трохи меншу кількість транзакцій.</li>
    <li>Порівняння з Yangon показує подібний тренд:</li>
        <ul style='font-family: Garamond ; font-size: 16px;'>
            <li>Учасники програм лояльності в обох містах демонструють вищі середні чеки.</li>
        </ul>
</ol> 

<p style='font-family: Garamond ; font-size: 16px;'>Для того, аби підвищити рівень умовної конверсії (from Normal to Member), слід розробити окрему стратегію залучення, окрім програм лояльності та бонусних акційних карток до покупок. Наприклад, можна створити оригінальну кампанію, що протягом місяця для новітніх Members будуть діяти окремі пропозиції щодо знижок на окремі групи товарів, за накопичені бонуси (придбання якогось товару у конкретній кількості) обмін на мерч компанії (супермаркету) - шопери, чашки, блокноти (різні види канцелярії); розробка конкурсів та рейтингів користувачів певних послуг супермаркетів (користування касою самообслуговування, придбання виключно екологічних біопакетів до множини покупок) тощо, для того, аби підсилити саме маркетингову стратегію транзитивного переходу від ординарного покупця до постійного, оскільки клієнту треба дати відчути себе особливим у моменті переходу від однієї категорії до іншої - скажімо, преміальної.</p>

<h1 style='font-family: Garamond ;'>Гіпотеза 3: Клієнти <b>Mandalay</b> мають кращий досвід покупок (user experience), аніж цільова аудиторія в <b>Yangon</b></h1>
<hr></hr>
<p style='font-family: Garamond ; font-size: 18px;'>Ефективність стратегії продажів і маркетингу (вищий <code>gross income</code>) в Mandalay підтверджується/спростовується розподілом оцінок користувачів, який буде проаналізовано за категоріями: низький (Low), середній (Medium) та високий (High). Розрахуємо відсоткове співвідношення кожної категорії, щоб оцінити, як розподіляються уподобання користувачів у Mandalay у порівнянні з Yangon. Цей аналіз надасть кількісну базу для оцінки тенденцій задоволеності клієнтів і їх відповідності маркетинговим стратегіям у кожному регіоні, а також надасть підстави рекомендувати впровадження нової бренд-стратегії мерченджайзингу.</p>

In [20]:
def category_mapping(value: Union[int, float]) -> str:
    """
    Map a numerical value to a categorical label based on predefined ranges.

    Parameters
    ----------
    value : Union[int, float]
        The numerical value to categorize.

    Returns
    -------
    str
        The categorical label for the given value. Possible values are:
        - 'High' for values >= 8
        - 'Medium' for values >= 6 and < 8
        - 'Low' for values < 6

    Examples
    --------
    >>> category_mapping(9)
    'High'

    >>> category_mapping(7)
    'Medium'

    >>> category_mapping(5)
    'Low'
    """
    if value >= 8:
        return 'High'
    elif value >= 6:
        return 'Medium'
    else:
        return 'Low'

In [21]:
filtered_data['Rating Category'] = filtered_data.Rating.apply(category_mapping)
filtered_data.head()

Invoice ID Branch    City Customer type  Gender            Product line  \
0  750-67-8428      A  Yangon        Member  Female       Health and beauty   
2  631-41-3108      A  Yangon        Normal    Male      Home and lifestyle   
3  123-19-1176      A  Yangon        Member    Male       Health and beauty   
4  373-73-7910      A  Yangon        Normal    Male       Sports and travel   
6  355-53-5943      A  Yangon        Member  Female  Electronic accessories   

   Unit price  Quantity   Tax 5%     Total       Date   Time      Payment  \
0       74.69         7  26.1415  548.9715   1/5/2019  13:08      Ewallet   
2       46.33         7  16.2155  340.5255   3/3/2019  13:23  Credit card   
3       58.22         8  23.2880  489.0480  1/27/2019  20:33      Ewallet   
4       86.31         7  30.2085  634.3785   2/8/2019  10:37      Ewallet   
6       68.84         6  20.6520  433.6920  2/25/2019  14:36      Ewallet   

     cogs  gross margin percentage  gross income  Rating Rating Category  
0  522.83                 4.761905       26.1415     9.1            High  
2  324.31                 4.761905       16.2155     7.4          Medium  
3  465.76                 4.761905       23.2880     8.4            High  
4  604.17                 4.761905       30.2085     5.3             Low  
6  413.04                 4.761905       20.6520     5.8             Low

In [22]:
filtered_data['Rating Category'] = filtered_data['Rating Category'].astype('category')

In [23]:
mandalay_rating_counts = filtered_data[filtered_data['City'] == 'Mandalay']
yangon_rating_counts = filtered_data[filtered_data['City'] == 'Yangon']

In [24]:
rating_counts_mandalay = mandalay_rating_counts["Rating Category"].value_counts()
rating_counts_mandalay/rating_counts_mandalay.sum() * 100

Rating Category
Medium    36.746988
Low       33.734940
High      29.518072
Name: count, dtype: float64

In [25]:
rating_counts_yangon = yangon_rating_counts["Rating Category"].value_counts()
rating_counts_yangon/rating_counts_yangon.sum() * 100

Rating Category
Medium    35.882353
High      32.941176
Low       31.176471
Name: count, dtype: float64

<ol style='font-family: Garamond ; font-size: 16px;'>
    <li>Попри те, що в Mandalay загальний середній чек вищий, розподіл оцінок користувачів не свідчить про кращий досвід покупок у цьому місті:</li>
        <ul ol style='font-family: Garamond ; font-size: 16px;'>
            <li>Частка високих оцінок у Yangon вища, ніж у Mandalay.</li>
            <li>Частка низьких оцінок у Mandalay також трохи більша.</li>
        </ul>
    <li>Це може означати, що:</li>
        <ul style='font-family: Garamond ; font-size: 16px;'>
            <li>У Mandalay ефективніші стратегії продажів можуть стимулювати більше покупок та вищі чеки, але не обов’язково призводять до кращого клієнтського досвіду.</li>
            <li>Є сенс дослідити додаткові фактори, такі як післяпродажне обслуговування, якість продукції чи зручність процесу покупки в Mandalay, щоб покращити задоволеність клієнтів.</li>
        </ul>
</ol>
<p style='font-family: Garamond ; font-size: 16px;'>Слід переглянути роботу сервісу та маркетингу з акцентом на покращення клієнтського досвіду. А також є необхідність впровадити нову бренд-стратегію мерчендайзингу, щоб зменшити частку низьких оцінок (visual merchandising, ексклюзивні пропозиції для Members, інтерактивні елементи для взаємодії з покупцями - дегустації, 3D цифрова реклама, storytelling у брендовому зонуванні, збір та реагування на фідбек прямо в точці продажу (QR-codes)).</p>

<h1 style='font-family: Garamond ;'>Гіпотеза 4: <b>Mandalay</b> має вищий обсяг покупок товарів із високою ціною (<code>unit price</code>)</h1>
<hr></hr>
<p style='font-family: Garamond ; font-size: 18px;'>Припустимо, що клієнти в Mandalay частіше купують дорожчі товари, що підвищує загальний gross income.</p>

In [30]:
unit_price_analysis = filtered_data.groupby('City').agg({
    'Unit price': ['mean', 'median', 'std'],
    'Total': ['sum', 'count']
}).reset_index()

unit_price_analysis.columns = ['City', 'Average Unit Price', 'Median Unit Price', 'Std Unit Price', 'Total Revenue', 'Transaction Count']

unit_price_analysis

City  Average Unit Price  Median Unit Price  Std Unit Price  \
0  Mandalay           55.659277             54.850       26.224559   
1    Yangon           54.780853             53.235       26.132127   

   Total Revenue  Transaction Count  
0    106197.6720                332  
1    106200.3705                340

<p style='font-family: Garamond ; font-size: 16px;'>У Mandalay середня та медіанна ціна товарів вищі, ніж у Yangon, що свідчить про те, що клієнти дійсно купують більш дорогі товари. Це частково пояснює вищий середній чек та gross income у Mandalay навіть при меншій кількості транзакцій.</p>
<p style='font-family: Garamond ; font-size: 16px;'>Також можна підкреслити, що приблизно однаковий Total Revenue досягається при меншій кількості транзакцій у Mandalay, тобто продаж кожного окремого товару там приносить більше доходу.</p>
<p style='font-family: Garamond ; font-size: 16px;'>Для Yangon можна запровадити множину таких рекомендацій згідно з наведеним вище аналізом щодо гіпотези 4:</p>
<ol style='font-family: Garamond ; font-size: 16px;'>
    <li>Запровадити крос-продажі та апсейлінг: пропонувати додаткові товари до покупки (комплекти, аксесуари).</li>
    <li>Запустити спеціальні акції на середньо- та високомаржинальні товари (наприклад, "купи товарів від $n$-ої суми — отримай бонус").</li>
    <li>Посилити маркетинг дорогих продуктів у Yangon через таргетовану рекламу та візуальні кампанії.</li>
</ol>

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [35]:
gross_margin_variance = filtered_data['gross margin percentage'].var()
gross_margin_variance

np.float64(0.0)

<p style='font-family: Garamond ; font-size: 18px;'><b>У всіх інвойсах однакова маржа</b>, оскільки міра розсіяння значень випадкової величини відносно середнього значення розподілу (variance) є нульовою. Більші значення дисперсії свідчать про більші відхилення значень випадкової величини від центру розподілу.</p>

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [83]:
product_line_distribution_mandalay = (
    filtered_data[filtered_data['City'] == 'Mandalay']
    .groupby('Product line')['Invoice ID']
    .count()
    .pipe(lambda x: x / x.sum())
)
product_line_distribution_mandalay.sort_values(ascending=False)

Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: Invoice ID, dtype: float64

In [84]:
product_line_distribution_yangon = (
    filtered_data[filtered_data['City'] == 'Yangon']
    .groupby('Product line')['Invoice ID']
    .count()
    .pipe(lambda x: x / x.sum())
)
product_line_distribution_yangon.sort_values(ascending=False)

Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: Invoice ID, dtype: float64